src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -587,440 +573,406 @@ private void dedups(ModuleDescriptor md) {
+                                               "entry",
+                                               desc,
+                                               true);
+                            cob.arrayStoreInstruction(TypeKind.ReferenceType);
+                            index++;
+                        }
 
-            // invoke Map.ofEntries(Map$Entry[])
-            mv.visitMethodInsn(INVOKESTATIC, "java/util/Map", "ofEntries",
-                    "([Ljava/util/Map$Entry;)Ljava/util/Map;", true);
-            mv.visitInsn(ARETURN);
-            mv.visitMaxs(0, 0);
-            mv.visitEnd();
+                        // invoke Map.ofEntries(Map$Entry[])
+                        cob.invokeInstruction(Opcode.INVOKESTATIC, CD_Map, "ofEntries",
+                                MethodTypeDesc.ofDescriptor(
+                                        "([Ljava/util/Map$Entry;)Ljava/util/Map;"), true);
+                        cob.returnInstruction(TypeKind.ReferenceType);
+                    }));
         }
 
         /**
          * Generate code to generate an immutable set.
          */
-        private void genImmutableSet(MethodVisitor mv, Set<String> set) {
+        private void genImmutableSet(CodeBuilder cob, Set<String> set) {
             int size = set.size();
 
             // use Set.of(Object[]) when there are more than 2 elements
             // use Set.of(Object) or Set.of(Object, Object) when fewer
             if (size > 2) {
-                pushInt(mv, size);
-                mv.visitTypeInsn(ANEWARRAY, "java/lang/String");
+                cob.constantInstruction(size);
+                cob.anewarray(CD_String);
                 int i = 0;
                 for (String element : sorted(set)) {
-                    mv.visitInsn(DUP);
-                    pushInt(mv, i);
-                    mv.visitLdcInsn(element);
-                    mv.visitInsn(AASTORE);
+                    cob.stackInstruction(Opcode.DUP);
+                    cob.constantInstruction(i);
+                    cob.constantInstruction(element);
+                    cob.arrayStoreInstruction(TypeKind.ReferenceType);
                     i++;
                 }
-                mv.visitMethodInsn(INVOKESTATIC,
-                        "java/util/Set",
+                cob.invokeInstruction(Opcode.INVOKESTATIC,
+                        CD_Set,
                         "of",
-                        "([Ljava/lang/Object;)Ljava/util/Set;",
+                        MethodTypeDesc.ofDescriptor("([Ljava/lang/Object;)Ljava/util/Set;"),
                         true);
             } else {
                 StringBuilder sb = new StringBuilder("(");
                 for (String element : sorted(set)) {
-                    mv.visitLdcInsn(element);
+                    cob.constantInstruction(element);
                     sb.append("Ljava/lang/Object;");
                 }
                 sb.append(")Ljava/util/Set;");
```

Comment:
The construction of the Set.of method type should move away from StringBuilder. It can be as:
```
var mtdArgs = new ClassDesc[size];
Arrays.fill(mtdArgs, CD_Object);
MethodTypeDesc mtd = MethodTypeDesc.of(CD_Set, mtdArgs);
```

Reply:
Applied, thanks.